File: ureka-framework-main/ureka_framework/view/menu_iot_device.py
# Environment
from ureka_framework.environment import Environment

# Resource (Logger)
from ureka_framework.resource.logger.simple_logger import simple_log

# Data Model (Message)
import json

# Data Model (RAM)
from ureka_framework.logic.device_controller import DeviceController
import ureka_framework.model.data_model.this_device as this_device

# Threading
import time


class MenuIoTDevice:
    ######################################################
    # Environment
    ######################################################
    @classmethod
    def set_environment(cls, situation: str = "measurement") -> None:
        if situation == "cli":
            Environment.DEPLOYMENT_ENV = "PRODUCTION"
            Environment.DEBUG_LOG = "CLOSED"
            Environment.CLI_LOG = "OPEN"
            Environment.MEASURE_LOG = "CLOSED"
            Environment.MORE_MEASURE_WORKER_LOG = "CLOSED"
            Environment.MORE_MEASURE_RESOURCE_LOG = "CLOSED"
        elif situation == "measurement":
            Environment.DEPLOYMENT_ENV = "PRODUCTION"
            Environment.DEBUG_LOG = "CLOSED"
            Environment.CLI_LOG = "CLOSED"
            Environment.MEASURE_LOG = "CLOSED"
            Environment.MORE_MEASURE_WORKER_LOG = "CLOSED"
            Environment.MORE_MEASURE_RESOURCE_LOG = "CLOSED"

    ######################################################
    # Secure Mode
    ######################################################
    def __init__(self, device_name: str) -> None:
        # GIVEN: Load IoTD
        self.iot_device = DeviceController(
            device_type=this_device.IOT_DEVICE,
            device_name=device_name,
        )
        # self.sleep_time = 0.2  # Simulate blocked I/O

    def get_iot_device(self) -> DeviceController:
        return self.iot_device

    def receive_u_ticket_through_bluetooth(self) -> DeviceController:
        # WHEN: Accept bluetooth connection from UA or CS
        Environment.COMMUNICATION_CHANNEL = "BLUETOOTH"
        self.iot_device.msg_receiver.accept_bluetooth_comm()

        # WHEN: Receive/Send Message in Connection
        self.iot_device.msg_receiver._recv_xxx_message()

        # RE-GIVEN: Close bluetooth Connection with UA or CS
        self.iot_device.msg_receiver.close_bluetooth_connection()

        # RE-GIVEN: Stop Accepting New bluetooth Connections from UA or CS
        self.iot_device.msg_receiver.close_bluetooth_acception()

        return self.iot_device

    ######################################################
    # Insecure Mode
    ######################################################
    def receive_insecure_cmd_through_bluetooth(
        self, option: str = "with_device_id"
    ) -> DeviceController:
        # WHEN: Accept bluetooth connection from UA or CS
        Environment.COMMUNICATION_CHANNEL = "BLUETOOTH"
        self.iot_device.msg_receiver.accept_bluetooth_comm()

        # WHEN: IoTD receive the insecure_cmd from UA or CS
        while True:
            try:
                # This will block until message is received
                insecure_cmd_json = (
                    self.iot_device.shared_data.connection_socket.recv_message()
                )

                ########################################################################
                # Message Size Measurement
                ########################################################################
                self.iot_device.measure_helper.measure_message_size(
                    "_recv_message", insecure_cmd_json
                )
                simple_log("cli", f"Received Command: {insecure_cmd_json}")

                ######################################################
                # Start Process Measurement
                ######################################################
                self.iot_device.measure_helper.measure_process_perf_start()

                # WHEN: IoTD do data processing
                if option == "shortest":
                    insecure_data_json = f"Data: {insecure_cmd_json}"
                    # if self.sleep_time > 0:
                    #     time.sleep(self.sleep_time)  # Simulate blocked I/O
                    #     self.sleep_time = self.sleep_time - 0.02
                else:
                    insecure_cmd_dict = json.loads(insecure_cmd_json)
                    insecure_data_dict = {
                        "protocol_verision": insecure_cmd_dict["protocol_verision"],
                        "device_id": insecure_cmd_dict["device_id"],
                        "insecure_data_response": f"Data: {insecure_cmd_dict['insecure_command']}",
                    }
                    insecure_data_json = json.dumps(insecure_data_dict, indent=4)
                    # if self.sleep_time > 0:
                    #     time.sleep(self.sleep_time)  # Simulate blocked I/O
                    #     self.sleep_time = self.sleep_time - 0.02

                # WHEN: IoTD return the insecure_data to UA or CS
                self.iot_device.shared_data.connection_socket.send_message(
                    insecure_data_json
                )
                simple_log("cli", f"Sent Data: {insecure_data_json}")

                ######################################################
                # End Process Measurement
                ######################################################
                self.iot_device.measure_helper.measure_recv_msg_perf_time(
                    "_device_recv_insecure_cmd"
                )

                ########################################################################
                # Start Comm Measurement
                ########################################################################
                if self.iot_device.shared_data.this_device.device_name != "iot_device":
                    self.iot_device.measure_helper.measure_comm_perf_start()

                simple_log(
                    "debug",
                    f"+ {self.iot_device.shared_data.this_device.device_name} manually finish CMD-DATA~~ (device)",
                )
            except OSError:
                simple_log("cli", f"")
                simple_log("cli", f"+ Connection is closed by peer.")
                break

        # RE-GIVEN: Close bluetooth connection with UA or CS
        self.iot_device.msg_receiver.close_bluetooth_connection()

        # RE-GIVEN: Stop Accepting New bluetooth Connections from UA or CS
        self.iot_device.msg_receiver.close_bluetooth_acception()

        return self.iot_device

File: ureka-framework-java/app/src/main/java/ureka/framework/view/MenuIoTDevice.java
package ureka.framework.view;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

import ureka.framework.Environment;
import ureka.framework.logic.DeviceController;
import ureka.framework.model.data_model.ThisDevice;
import ureka.framework.resource.crypto.SerializationUtil;
import ureka.framework.resource.logger.SimpleLogger;

public class MenuIoTDevice {
    // Environment
    public static void setEnvironment(String situation) {
        if (situation.equals("cli")) {
            Environment.DEPLOYMENT_ENV = "PRODUCTION";
            Environment.DEBUG_LOG = "CLOSED";
            Environment.CLI_LOG = "OPEN";
            Environment.MEASURE_LOG = "CLOSED";
            Environment.MORE_MEASURE_WORKER_LOG = "CLOSED";
            Environment.MORE_MEASURE_RESOURCE_LOG = "CLOSED";
        } else if (situation.equals("measurement")) {
            Environment.DEPLOYMENT_ENV = "PRODUCTION";
            Environment.DEBUG_LOG = "CLOSED";
            Environment.CLI_LOG = "CLOSED";
            Environment.MEASURE_LOG = "CLOSED";
            Environment.MORE_MEASURE_WORKER_LOG = "CLOSED";
            Environment.MORE_MEASURE_RESOURCE_LOG = "CLOSED";
        }
    }

    private DeviceController iotDevice;

    // Secure Mode
    public MenuIoTDevice(String deviceName) {
        // GIVEN: Load IoTD
        this.iotDevice = new DeviceController(ThisDevice.IOT_DEVICE,deviceName);
    }

    public DeviceController getIotDevice() {
        return this.iotDevice;
    }

    public DeviceController receiveUTicketThroughBluetooth() {
        // WHEN: Accept bluetooth connection from UA or CS
        Environment.COMMUNICATION_CHANNEL = "BLUETOOTH";
//        this.iotDevice.getMsgReceiver().acceptBluetoothComm();

        // WHEN: Receive/Send Message in Connection
        this.iotDevice.getMsgReceiver()._recvXxxMessage();

        // RE-GIVEN: Close bluetooth Connection with UA or CS
//        this.iotDevice.getMsgReceiver().closeBluetoothConnection();

        // RE-GIVEN: Stop Accepting New bluetooth Connections from UA or CS
//        this.iotDevice.getMsgReceiver().closeBluetoothAcception();
        return this.iotDevice;
    }

    // Insecure Mode
    public DeviceController receiveInsecureCmdThroughBluetooth(String option) {
        // WHEN: Accept bluetooth connection from UA or CS
        Environment.COMMUNICATION_CHANNEL = "BLUETOOTH";
//        this.iotDevice.getMsgReceiver().acceptBluetoothComm();

        // WHEN: IoTD receive the insecure_cmd from UA or CS
        while (true) {
            try {
                // This will block until message is received
                String insecureCmdJson = "";
//                insecureCmdJson = this.iotDevice.getSharedData().getConnectionSocket().recvMessage();

                // Message Size Measurement
                this.iotDevice.getMeasureHelper().measureMessageSize("_recvMessage",insecureCmdJson);
                SimpleLogger.simpleLog("cli", "Received Command: " + insecureCmdJson);

                // Start Process Measurement
                this.iotDevice.getMeasureHelper().measureProcessPerfStart();

                // WHEN: IoTD do data processing
                String insecureDataJson;
                if (option.equals("shortest")) {
                    insecureDataJson = "Data: " + insecureCmdJson;
                } else {
                    Map<String, String> insecureCmdDict = SerializationUtil.jsonStrToDict(insecureCmdJson);
                    Map<String, String> insecureDataDict = new HashMap<>();
                    insecureDataDict.put("protocolVersion", insecureCmdDict.get("protocolVersion"));
                    insecureDataDict.put("deviceId", insecureCmdDict.get("deviceId"));
                    insecureDataDict.put("insecureDataResponse", "Data: " + insecureCmdDict.get("insecureCommand"));

                    insecureDataJson = SerializationUtil.dictToJsonStr(insecureDataDict);
                }
                // WHEN: IoTD return the insecure_data to UA or CS
//                this.iotDevice.getSharedData().getConnectionSocket().sendMessage(insecureDataJson);
                SimpleLogger.simpleLog("cli","Sent Data: " + insecureDataJson);

                // End Process Measurement
                this.iotDevice.getMeasureHelper().measureRecvMsgPerfTime("_deviceRecvInsecureCmd");

                // Start Comm Measurement
                if(!this.iotDevice.getSharedData().getThisDevice().getDeviceName().equals("iotDevice")) {
                    this.iotDevice.getMeasureHelper().measureCommPerfStart();
                    SimpleLogger.simpleLog("debug", this.iotDevice.getSharedData().getThisDevice().getDeviceName() + " manually finish CMD-DATA~~ (device)");
                }
//            } catch (IOException e) {
//                SimpleLogger.simpleLog("cli", "");
//                SimpleLogger.simpleLog("cli", "+ Connection is closed by peer.");
//                break;
            } catch (RuntimeException e) {
                SimpleLogger.simpleLog("cli", "");
                SimpleLogger.simpleLog("cli", "+ Connection is closed by peer.");
                break;
            }
        }
        // RE-GIVEN: Close bluetooth connection with UA or CS
//        this.iotDevice.getMsgReceiver().closeBluetoothConnection();
        // RE-GIVEN: Stop Accepting New bluetooth Connections from UA or CS
//        this.iotDevice.getMsgReceiver().closeBluetoothAcception();
        return this.iotDevice;
    }
}
